src/algebra/group/prod.lean

```diff
@@ -0,0 +1,236 @@
+@[to_additive "Given additive monoids `A`, `B`, the natural inclusion homomorphism
+from `A` to `A × B`."]
+def inl : M →* M × N :=
+⟨λ x, (x, 1), rfl, λ _ _, prod.ext rfl (one_mul 1).symm⟩
+
+/-- Given monoids `M`, `N`, the natural inclusion homomorphism from `N` to `M × N`. -/
+@[to_additive "Given additive monoids `A`, `B`, the natural inclusion homomorphism
+from `B` to `A × B`."]
+def inr : N →* M × N :=
+⟨λ y, (1, y), rfl, λ _ _, prod.ext (one_mul 1).symm rfl⟩
+
+variables {M N}
+
+@[to_additive] lemma coe_fst : ⇑(fst M N) = prod.fst := rfl
+@[to_additive] lemma coe_snd : ⇑(snd M N) = prod.snd := rfl
+
+@[to_additive] lemma inl_apply (x) : inl M N x = (x, 1) := rfl
+@[to_additive] lemma inr_apply (y) : inr M N y = (1, y) := rfl
+
+@[simp, to_additive] lemma fst_comp_inl : (fst M N).comp (inl M N) = id M := rfl
+@[simp, to_additive] lemma snd_comp_inl : (snd M N).comp (inl M N) = 1 := rfl
+@[simp, to_additive] lemma fst_comp_inr : (fst M N).comp (inr M N) = 1 := rfl
+@[simp, to_additive] lemma snd_comp_inr : (snd M N).comp (inr M N) = id N := rfl
+
+@[simp, to_additive] lemma fst_inl (x) : (inl M N x).1 = x := rfl
+@[simp, to_additive] lemma snd_inl (x) : (inl M N x).2 = 1 := rfl
+@[simp, to_additive] lemma fst_inr (y) : (inr M N y).1 = 1 := rfl
+@[simp, to_additive] lemma snd_inr (y) : (inr M N y).2 = y := rfl
+
+section prod
+
+variable [monoid P]
+
+/-- Combine two `monoid_hom`s `f : M →* N`, `g : M →* P` into `f.prod g : M →* N × P`
+given by `(f.prod g) x = (f x, g x)` -/
+@[to_additive prod "Combine two `add_monoid_hom`s `f : M →+ N`, `g : M →+ P` into
+`f.prod g : M →+ N × P` given by `(f.prod g) x = (f x, g x)`"]
+def prod (f : M →* N) (g : M →* P) : M →* N × P :=
+{ to_fun := λ x, (f x, g x),
+  map_one' := prod.ext f.map_one g.map_one,
+  map_mul' := λ x y, prod.ext (f.map_mul x y) (g.map_mul x y) }
+
+@[to_additive prod_apply]
+lemma prod_apply (f : M →* N) (g : M →* P) (x) : f.prod g x = (f x, g x) := rfl
+
+@[to_additive fst_compm_prod]
+lemma fst_comp_prod (f : M →* N) (g : M →* P) : (fst N P).comp (f.prod g) = f :=
+ext $ λ x, rfl
+
+@[to_additive snd_comp_prod]
+lemma snd_comp_prod (f : M →* N) (g : M →* P) : (snd N P).comp (f.prod g) = g :=
+ext $ λ x, rfl
+
+@[to_additive prod_unique]
+lemma prod_unique (f : M →* N × P) :
+  ((fst N P).comp f).prod ((snd N P).comp f) = f :=
+ext $ λ x, by simp only [prod_apply, coe_fst, coe_snd, comp_apply, prod.mk.eta]
+
+end prod
+
+section prod_map
+
+variables {M' : Type*} {N' : Type*} [monoid M'] [monoid N'] [monoid P]
+  (f : M →* M') (g : N →* N')
+
+/-- `prod.map` as a `monoid_hom`. -/
+@[to_additive prod_map "`prod.map` as an `add_monoid_hom`"]
+def prod_map : M × N →* M' × N' := (f.comp (fst M N)).prod (g.comp (snd M N))
+
+@[to_additive prod_map_def]
+lemma prod_map_def : prod_map f g = (f.comp (fst M N)).prod (g.comp (snd M N)) := rfl
+
+-- TODO : use `rfl` once we redefine `prod.map` in stdlib
+@[to_additive coe_prod_map]
```

Comment:
```suggestion
@[simp, to_additive coe_prod_map]
```
(or provide the simp lemma for the coercion applied to arguments; I'm not sure if it makes much difference, but I want one or the other)